promql/parser/generated_parser.y

```diff
@@ -436,19 +436,27 @@ paren_expr      : LEFT_PAREN expr RIGHT_PAREN
 offset_expr: expr OFFSET number_duration_literal
                         {
   		            numLit, _ := $3.(*NumberLiteral)
-      		            dur := time.Duration(numLit.Val * 1000) * time.Millisecond
+                            dur := &NumberLiteral{Val: numLit.Val}
                  	    yylex.(*parser).addOffset($1, dur)
                             $$ = $1
                         }
                 | expr OFFSET SUB number_duration_literal
                         {
 			    numLit, _ := $4.(*NumberLiteral)
-		            dur := time.Duration(numLit.Val * 1000) * time.Millisecond
-			    yylex.(*parser).addOffset($1, -dur)
+                            dur := &NumberLiteral{Val: numLit.Val * -1.0}
+			    yylex.(*parser).addOffset($1, dur)
                             $$ = $1
                         }
+                | expr OFFSET paren_expr
+                        {
+                            dur, _ := $3.(*ParenExpr)
+                            if dur.Type() != ValueTypeScalar {
+                                yylex.(*parser).unexpected("offset", "time expression does not evaluate to a scalar")
+                            }
+                            yylex.(*parser).addOffset($1, dur)
+                        }
                 | expr OFFSET error
-                        { yylex.(*parser).unexpected("offset", "number or duration"); $$ = $1 }
+                        { yylex.(*parser).unexpected("offset", "number or duration or time expression in parentheses"); $$ = $1 }
```

Comment:
I guess we should explain "time expression" when documenting this feature. (Can be used for timestamps as well as durations, has to be scalar, must not touch TSDB, must not depend on the eval time.)